src/main/kotlin/nl/nuts/discovery/service/CertificateAndKeyService.kt

```diff
@@ -19,50 +19,221 @@
+import org.bouncycastle.asn1.x500.style.BCStyle
+import org.bouncycastle.asn1.x509.Extension
+import org.bouncycastle.asn1.x509.GeneralName
+import org.bouncycastle.asn1.x509.GeneralNames
+import org.bouncycastle.operator.jcajce.JcaContentVerifierProviderBuilder
+import org.bouncycastle.pkcs.jcajce.JcaPKCS10CertificationRequest
+import org.bouncycastle.util.io.pem.PemReader
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+import java.io.File
+import java.io.IOException
+import java.io.Reader
+import java.net.URISyntaxException
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.security.GeneralSecurityException
+import java.security.KeyFactory
+import java.security.KeyPair
+import java.security.PrivateKey
 import java.security.cert.X509Certificate
+import java.security.spec.PKCS8EncodedKeySpec
+import java.util.*
+import javax.transaction.Transactional
 
-interface CertificateAndKeyService {
+
+/**
+ * Certificate and key logic
+ */
+//@Profile(value = arrayOf("dev", "test", "default"))
+@Service
+class CertificateAndKeyService {
+
+    @Autowired
+    lateinit var nutsDiscoveryProperties: NutsDiscoveryProperties
+
+    @Autowired
+    lateinit var certificateRepository: CertificateRepository
+
+    @Autowired
+    lateinit var certificateRequestRepository: CertificateRequestRepository
 
     /**
-     * Sign a certificate based on the name
+     * Create a certificate, add the email extension with email from the request
+     * and sign it with the Nuts intermediate keyPair.
+     *
+     * Deletes request and stores certificate
      */
-    fun signCertificate(request: PKCS10CertificationRequest): X509Certificate
+    @Transactional
+    fun signCertificate(request: CertificateRequest): X509Certificate {
+        val pkcs10 = JcaPKCS10CertificationRequest(request.toPKCS10())
+        val name = CordaX500Name.parse(request.name!!)
+
+        val issuerCertificate = intermediateCertificate()
+        val certificateType = CertificateType.NODE_CA
+        val issuer = issuerCertificate.subjectX500Principal
+        val issuerKeyPair = intermediateKeyPair()
+        val subject = name.x500Principal
+        val subjectPublicKey = pkcs10.publicKey
+        val validityWindow = X509Utilities.DEFAULT_VALIDITY_WINDOW
+        val window = X509Utilities.getCertificateValidityWindow(validityWindow.first, validityWindow.second, issuerCertificate)
+
+        val signatureScheme = Crypto.findSignatureScheme(issuerKeyPair.private)
+        val provider = Crypto.findProvider(signatureScheme.providerName)
+        val signer = ContentSignerBuilder.build(signatureScheme, issuerKeyPair.private, provider)
+        val builder = X509Utilities.createPartialCertificate(
+            certificateType,
+            issuer,
+            issuerKeyPair.public,
+            subject,
+            subjectPublicKey,
+            window,
+            null, // todo
```

Comment:
TODO?

Reply:
Nuts CA tree, constraints like in registry. I added some text to the todo